File: Core/src/main/java/com/github/ezauton/core/utils/Stopwatch.java
package com.github.ezauton.core.utils;

import java.util.concurrent.TimeUnit;

/**
 * A handy stopwatch for recording time in seconds every time it is polled
 */
public class Stopwatch
{

    protected final IClock clock;
    protected long millis = -1;

    public Stopwatch(IClock clock)
    {
        this.clock = clock;
    }

    public void init()
    {
        millis = clock.getTime();

    }

    /**
     * Read and reset
     *
     * @return The value of the stopwatch (ms)
     */
    public double pop()
    {
        double readVal = read();
        reset();
        return readVal;
    }

    /**
     * Read and reset
     *
     * @param timeUnit The time unit you would like to get the result in
     * @return Value of stopwatch (in specified timeunit)
     */
    public double pop(TimeUnit timeUnit)
    {
        return pop() / timeUnit.toMillis(1);
    }

    public IClock getClock()
    {
        return clock;
    }

    /**
     * Read without resetting
     *
     * @return The value of the stopwatch (ms)
     */
    public long read()
    {
        if(!isInit()) throw new IllegalArgumentException("Stopwatch must be initialized to use");
        return clock.getTime() - millis;
    }

    public long read(TimeUnit timeUnit)
    {
        return timeUnit.convert(read(), TimeUnit.MILLISECONDS);
    }

    /**
     * Reset without reading
     */
    public Stopwatch reset()
    {
        millis = clock.getTime();
        return this;
    }

    /**
     * @return If this stopwatch is initialized
     */
    public boolean isInit()
    {
        return millis != -1;
    }

    /**
     * @return If is not init
     */
    public boolean resetIfNotInit()
    {
        if(isInit())
        {
            return false;
        }
        reset();
        return true;
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/utils/Stopwatch.kt
package com.github.ezauton.core.utils

import com.github.ezauton.conversion.Time
import com.github.ezauton.conversion.now
import com.github.ezauton.conversion.seconds

/**
 * A handy stopwatch for recording time in seconds since it was last polled. Requires a ⏱ [Clock] to keep track
 * of time.
 */
class Stopwatch {
  companion object {
    fun start(): Stopwatch {
      return Stopwatch().apply { init() }
    }
  }
  private var startDuration: Time = (-1).seconds

  /**
   * @return If this stopwatch is initialized
   */
  val isInit: Boolean get() = startDuration.value >= 0.0


  fun init() {
    startDuration = now()
  }

  /**
   * Read and reset
   *
   * @return The value of the stopwatch (ms)
   */
  fun pop(): Time {
    val readVal = read()
    reset()
    return readVal
  }

  /**
   * Read without resetting
   *
   * @return The value of the stopwatch (ms)
   */
  fun read(): Time {
    if (!isInit) throw IllegalArgumentException("Stopwatch must be initialized to use")
    return now() - startDuration
  }

  /**
   * Reset without reading
   */
  fun reset(): Stopwatch {
    startDuration = now()
    return this
  }

  /**
   * @return If is not init
   */
  fun resetIfNotInit(): Boolean {
    if (isInit) {
      return false
    }
    reset()
    return true
  }
}
